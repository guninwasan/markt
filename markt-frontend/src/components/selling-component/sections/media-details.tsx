import React, { useRef } from "react";
import {
  FormGroup,
  Label,
  AddMediaButton,
  SectionHeader,
  PreviewImagesContainer,
  RemoveButton,
  StyledImage,
  StyledVideo,
} from "../selling-component.styles";

type MediaSectionProps = {
  mediaFiles: File[];
  setMediaFiles: React.Dispatch<React.SetStateAction<File[]>>;
  displayImage: File | null;
  setDisplayImage: React.Dispatch<React.SetStateAction<File | null>>;
};

const maxMediaFiles = 4;

const MediaSection = ({
  mediaFiles,
  setMediaFiles,
  displayImage,
  setDisplayImage,
}: MediaSectionProps) => {
  const displayImageInputRef = useRef<HTMLInputElement>(null);
  const mediaInputRef = useRef<HTMLInputElement>(null);

  const addDisplayImage = (event: React.ChangeEvent<HTMLInputElement>) => {
    const files = event.target.files ? Array.from(event.target.files) : [];
    if (files.length > 0) {
      if (files[0].size > 1024 * 1024) {
        alert("Image size should not exceed 1MB.");
        return;
      }
      setDisplayImage(files[0]);
    }
  };

  const handleMediaChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const files = Array.from(event.target.files || []);
    const validFiles = files.filter((file) => {
      if (file.size > 1024 * 1024) {
        alert(`File ${file.name} exceeds the 1MB size limit.`);
        return false;
      }
      return true;
    });

    if (mediaFiles.length + validFiles.length > maxMediaFiles) {
      alert(`You can only upload up to ${maxMediaFiles} files.`);
      return;
    }

    setMediaFiles((prevMediaFiles) => [...prevMediaFiles, ...validFiles]);
  };

  const handleRemoveDisplay = () => {
    if (window.confirm("Are you sure you want to remove the display image?")) {
      setDisplayImage(null);
    }
  };

  const handleRemoveMedia = (index: number) => {
    if (window.confirm("Are you sure you want to remove this media file?")) {
      setMediaFiles((prevMediaFiles) => [
        ...prevMediaFiles.slice(0, index),
        ...prevMediaFiles.slice(index + 1),
      ]);
    }
  };

  return (
    <>
      <SectionHeader>Media</SectionHeader>
      <FormGroup>
        <Label htmlFor="displayImage">Display Image *</Label>
        <AddMediaButton
          type="button"
          onClick={() => displayImageInputRef.current?.click()}
        >
          Add Display Image (Max 1 MB) *
        </AddMediaButton>
        <input
          id="displayImageInput"
          type="file"
          accept="image/*"
          onChange={addDisplayImage}
          style={{ display: "none" }}
          ref={displayImageInputRef}
          data-testid="displayImageInput"
        />
        {displayImage && (
          <PreviewImagesContainer>
            <StyledImage
              src={URL.createObjectURL(displayImage)}
              alt="display-preview"
            />
            <RemoveButton onClick={handleRemoveDisplay}>✕</RemoveButton>
          </PreviewImagesContainer>
        )}
        <br />
        <Label htmlFor="mediaInput">
          Additional Images and Videos (Max {maxMediaFiles} )
        </Label>
        <AddMediaButton
          type="button"
          onClick={() => mediaInputRef.current?.click()}
        >
          + Add Images or Videos (Max 1 MB)
        </AddMediaButton>
        <input
          id="mediaInput"
          type="file"
          multiple
          accept="image/*,video/*"
          onChange={handleMediaChange}
          style={{ display: "none" }}
          ref={mediaInputRef}
          data-testid="mediaInput"
        />
        <p>
          {mediaFiles.length} / {maxMediaFiles} files uploaded
        </p>
        {mediaFiles.map((file, index) => (
          <PreviewImagesContainer key={index}>
            {file.type.startsWith("image/") ? (
              <StyledImage
                src={URL.createObjectURL(file)}
                alt={`media-preview-${index + 1}`}
              />
            ) : file.type.startsWith("video/") ? (
              <StyledVideo controls>
                <source src={URL.createObjectURL(file)} type={file.type} />
              </StyledVideo>
            ) : null}
            <RemoveButton onClick={() => handleRemoveMedia(index)}>
              ✕
            </RemoveButton>
          </PreviewImagesContainer>
        ))}
      </FormGroup>
    </>
  );
};

export { MediaSection };
