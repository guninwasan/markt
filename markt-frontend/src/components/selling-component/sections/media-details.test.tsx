import React from "react";
import { render, fireEvent, screen, waitFor } from "@testing-library/react";
import { MediaSection } from "./media-details";

const createFile = (name: string, type: string, sizeInKB: number): File => {
  return new File([new Array(sizeInKB * 1024).fill("a").join("")], name, {
    type,
  });
};

const createImageFile = (sizeInKB: number): File => {
  return createFile("test-image.jpg", "image/jpeg", sizeInKB);
};

describe("MediaSection Component", () => {
  const mockSetMediaFiles = jest.fn();
  const mockSetDisplayImage = jest.fn();
  const maxMediaFiles = 4;

  beforeAll(() => {
    global.URL.createObjectURL = jest.fn(() => "mocked-url");
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  const renderComponent = () =>
    render(
      <MediaSection
        mediaFiles={[]}
        setMediaFiles={mockSetMediaFiles}
        displayImage={null}
        setDisplayImage={mockSetDisplayImage}
      />
    );

  it("should render component correctly", () => {
    renderComponent();
    expect(screen.getByText(/Media/)).toBeInTheDocument();
    expect(screen.getByText(/Add Display Image/)).toBeInTheDocument();
    expect(screen.getByText(/Add Images or Videos/)).toBeInTheDocument();
  });

  it("should add display image if file size is within limit", async () => {
    renderComponent();
    const file = createImageFile(500);
    const input = screen.getByTestId("displayImageInput");
    fireEvent.change(input, { target: { files: [file] } });

    await waitFor(() => {
      expect(mockSetDisplayImage).toHaveBeenCalledWith(file);
    });
  });

  it("should show alert if display image exceeds 1 MB", () => {
    renderComponent();
    const file = createImageFile(1500);
    const input = screen.getByTestId("displayImageInput");
    window.alert = jest.fn();
    fireEvent.change(input, { target: { files: [file] } });

    expect(window.alert).toHaveBeenCalledWith(
      "Image size should not exceed 1MB."
    );
    expect(mockSetDisplayImage).not.toHaveBeenCalled();
  });

  it("should add media file if file size is within limit", async () => {
    renderComponent();
    const file = createImageFile(500);
    const input = screen.getByTestId("mediaInput");
    fireEvent.change(input, { target: { files: [file] } });

    await waitFor(() => {
      expect(mockSetMediaFiles).toHaveBeenCalledWith(expect.any(Function));
    });
  });

  it("should show alert if adding media files exceeds max limit", () => {
    const existingFiles = Array.from({ length: maxMediaFiles }, () =>
      createImageFile(500)
    );
    render(
      <MediaSection
        mediaFiles={existingFiles}
        setMediaFiles={mockSetMediaFiles}
        displayImage={null}
        setDisplayImage={mockSetDisplayImage}
      />
    );

    const file = createImageFile(500);
    const input = screen.getByTestId("mediaInput");
    window.alert = jest.fn();
    fireEvent.change(input, { target: { files: [file] } });

    expect(window.alert).toHaveBeenCalledWith(
      `You can only upload up to ${maxMediaFiles} files.`
    );
    expect(mockSetMediaFiles).not.toHaveBeenCalled();
  });

  it("should remove display image after confirmation", () => {
    const displayImage = createImageFile(500);

    render(
      <MediaSection
        mediaFiles={[]}
        setMediaFiles={mockSetMediaFiles}
        displayImage={displayImage}
        setDisplayImage={mockSetDisplayImage}
      />
    );

    window.confirm = jest.fn(() => true);
    fireEvent.click(screen.getByText("âœ•"));

    expect(window.confirm).toHaveBeenCalledWith(
      "Are you sure you want to remove the display image?"
    );
    expect(mockSetDisplayImage).toHaveBeenCalledWith(null);
  });

  it("should match snapshot", () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });
});
