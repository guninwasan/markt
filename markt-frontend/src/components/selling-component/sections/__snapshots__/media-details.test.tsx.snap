// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MediaSection Component should match snapshot 1`] = `
<div>
  <h2
    class="sc-gtLWhw eYVYUD"
  >
    Media
  </h2>
  <div
    class="sc-egkSDF bHeElY"
  >
    <label
      class="sc-dntaoT bUVQgQ"
      for="displayImage"
    >
      Display Image *
    </label>
    <button
      class="sc-kLhKbu dwqFqP"
      type="button"
    >
      Add Display Image (Max 1 MB)
    </button>
    <input
      accept="image/*"
      data-testid="displayImageInput"
      id="displayImageInput"
      style="display: none;"
      type="file"
    />
    <br />
    <label
      class="sc-dntaoT bUVQgQ"
      for="mediaInput"
    >
      Additional Images and Videos (Max 
      4
       )
    </label>
    <button
      class="sc-kLhKbu dwqFqP"
      type="button"
    >
      + Add Images or Videos (Max 1 MB)
    </button>
    <input
      accept="image/*,video/*"
      data-testid="mediaInput"
      id="mediaInput"
      multiple=""
      style="display: none;"
      type="file"
    />
    <p>
      0
       / 
      4
       files uploaded
    </p>
  </div>
</div>
`;
